### PR TITLE
Fix issue #920 - Fixed endianess on netlink message parsing

### PR DESCRIPTION
**Pull request by uskr**
I had to redo this PR due to a missing file.

### CODE REVIEW

**Comment by lennartkoopmann** [comment]
The initialization fails for me:

```
[INFO][2023-09-05 20:50:44][brokers::dot11_broker] Installing WiFi broker thread <0>.
[INFO][2023-09-05 20:50:44][brokers::ethernet_broker] Installing ethernet broker thread <0>.
[INFO][2023-09-05 20:50:44][ethernet::capture] Starting ethernet capture on [enp6s0].
[INFO][2023-09-05 20:50:44][dot11::capture] Starting WiFi capture on [wlx00c0ca956856]
[INFO][2023-09-05 20:50:44][dot11::capture] Starting WiFi capture on [wlx00c0ca971215]
[INFO][2023-09-05 20:50:44][dot11::capture] Temporarily disabling interface [wlx00c0ca971215] ...
[INFO][2023-09-05 20:50:44][dot11::capture] Temporarily disabling interface [wlx00c0ca956856] ...
[ERROR][2023-09-05 20:50:44][dot11::capture] Could not disable device [wlx00c0ca971215]: Could not load device with name [wlx00c0ca971215] for changing state: Could not fetch devices: Could not parse GetIf Netlink response: Missing necessary attributes
[ERROR][2023-09-05 20:50:44][nzyme_tap] WiFi capture [wlx00c0ca971215] disconnected. Retrying in 5 seconds.
[ERROR][2023-09-05 20:50:44][dot11::capture] Could not disable device [wlx00c0ca956856]: Could not load device with name [wlx00c0ca956856] for changing state: Could not fetch devices: Could not parse GetIf Netlink response: Missing necessary attributes
[ERROR][2023-09-05 20:50:44][nzyme_tap] WiFi capture [wlx00c0ca956856] disconnected. Retrying in 5 seconds.
[ERROR][2023-09-05 20:50:44][dot11::channel_hopper] Could not fetch information of device [wlx00c0ca956856]. Not assigning to channels: Could not fetch devices: Could not parse GetIf Netlink response: Missing necessary attributes
[ERROR][2023-09-05 20:50:44][dot11::channel_hopper] Could not fetch information of device [wlx00c0ca971215]. Not assigning to channels: Could not fetch devices: Could not parse GetIf Netlink response: Missing necessary attributes
[INFO][2023-09-05 20:50:44][dot11::channel_hopper] Available adapters and channels: {}
[ERROR][2023-09-05 20:50:44][nzyme_tap] Could not initialize ChannelHopper: WiFi adapter [wlx00c0ca956856] not found.
```

What information should I provide for you to debug?

**Comment by lennartkoopmann** [comment]
Attaching the complete output of `iw phy`:

[iwphy.txt](https://github.com/nzymedefense/nzyme/files/12530404/iwphy.txt)

**Comment by uskr** [comment]
Thanks, let me check on that.

**Comment by uskr** [comment]
Can you do a "ip link show" or "ifconfig -a" and post here, please.

**Comment by uskr** [comment]
also kernel version, please. uname -r is enough.

**Comment by lennartkoopmann** [comment]
For sure:

```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: enp6s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP mode DEFAULT group default qlen 1000
    link/ether 74:56:3c:43:77:af brd ff:ff:ff:ff:ff:ff
3: wlo1: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN mode DORMANT group default qlen 1000
    link/ether f4:c8:8a:2e:df:74 brd ff:ff:ff:ff:ff:ff
    altname wlp0s20f3
4: wlx00c0ca971215: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/ieee802.11/radiotap 00:c0:ca:97:12:15 brd ff:ff:ff:ff:ff:ff
5: wlx00c0ca956856: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/ieee802.11/radiotap 00:c0:ca:95:68:56 brd ff:ff:ff:ff:ff:ff
6: tailscale0: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1280 qdisc pfifo_fast state UNKNOWN mode DEFAULT group default qlen 500
    link/none 
```

```
enp6s0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 172.16.0.140  netmask 255.255.255.0  broadcast 172.16.0.255
        inet6 fe80::5782:158e:b66d:ab75  prefixlen 64  scopeid 0x20<link>
        ether 74:56:3c:43:77:af  txqueuelen 1000  (Ethernet)
        RX packets 4305846  bytes 4379558776 (4.3 GB)
        RX errors 0  dropped 298  overruns 0  frame 0
        TX packets 1461419  bytes 186894582 (186.8 MB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 85272799  bytes 18353778608 (18.3 GB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 85272799  bytes 18353778608 (18.3 GB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

tailscale0: flags=4305<UP,POINTOPOINT,RUNNING,NOARP,MULTICAST>  mtu 1280
        inet 100.87.54.63  netmask 255.255.255.255  destination 100.87.54.63
        inet6 fe80::bccf:f14a:fa9a:740c  prefixlen 64  scopeid 0x20<link>
        inet6 fd7a:115c:a1e0:ab12:4843:cd96:6257:363f  prefixlen 128  scopeid 0x0<global>
        unspec 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  txqueuelen 500  (UNSPEC)
        RX packets 2890  bytes 2519422 (2.5 MB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 20902  bytes 3971607 (3.9 MB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

wlo1: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        ether f4:c8:8a:2e:df:74  txqueuelen 1000  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

wlx00c0ca956856: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        unspec 00-C0-CA-95-68-56-00-AE-00-00-00-00-00-00-00-00  txqueuelen 1000  (UNSPEC)
        RX packets 2453152  bytes 264874033 (264.8 MB)
        RX errors 0  dropped 3050  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

wlx00c0ca971215: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        unspec 00-C0-CA-97-12-15-00-AE-00-00-00-00-00-00-00-00  txqueuelen 1000  (UNSPEC)
        RX packets 1879934  bytes 391336912 (391.3 MB)
        RX errors 0  dropped 2880  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

```
6.2.0-31-generic
```

**Comment by uskr** [comment]
Can you run the tap on TRACE log level and attach the output here, please sir. I should be able to at least pinpoint the issue with that.

**Comment by uskr** [comment]
can you checkout the last version of the 920-endianess-fix branch and see how it goes?

The new code already support wifi 6e. I will do a through explanation of what was happening and what I had to do.

**Comment by lennartkoopmann** [comment]
Confirmed that I'm on the latest revision, but still getting the same errors:

```
[INFO][2023-09-08 19:18:38][logging] Initialized logging with level filter [INFO].
[INFO][2023-09-08 19:18:38][nzyme_tap] Starting nzyme tap version [2.0.0-alpha.4-SNAPSHOT].
[INFO][2023-09-08 19:18:38][nzyme_tap] Parsed and loaded configuration.
[INFO][2023-09-08 19:18:38][ethernet::capture] Available network device: enp6s0 (no description): [172.16.0.140, fe80::5782:158e:b66d:ab75]
[INFO][2023-09-08 19:18:38][ethernet::capture] Available network device: wlx00c0ca971215 (no description): [no interface addresses]
[INFO][2023-09-08 19:18:38][ethernet::capture] Available network device: wlx00c0ca956856 (no description): [no interface addresses]
[INFO][2023-09-08 19:18:38][ethernet::capture] Available network device: tailscale0 (no description): [100.87.54.63, fd7a:115c:a1e0:ab12:4843:cd96:6257:363f, fe80::bccf:f14a:fa9a:740c]
[INFO][2023-09-08 19:18:38][ethernet::capture] Available network device: any (Pseudo-device that captures on all interfaces): [no interface addresses]
[INFO][2023-09-08 19:18:38][ethernet::capture] Available network device: lo (no description): [127.0.0.1, ::1]
[INFO][2023-09-08 19:18:38][ethernet::capture] Available network device: wlo1 (no description): [no interface addresses]
[INFO][2023-09-08 19:18:38][ethernet::capture] Available network device: bluetooth0 (Bluetooth adapter number 0): [no interface addresses]
[INFO][2023-09-08 19:18:38][ethernet::capture] Available network device: bluetooth-monitor (Bluetooth Linux Monitor): [no interface addresses]
[INFO][2023-09-08 19:18:38][ethernet::capture] Available network device: nflog (Linux netfilter log (NFLOG) interface): [no interface addresses]
[INFO][2023-09-08 19:18:38][ethernet::capture] Available network device: nfqueue (Linux netfilter queue (NFQUEUE) interface): [no interface addresses]
[INFO][2023-09-08 19:18:38][ethernet::capture] Available network device: dbus-system (D-Bus system bus): [no interface addresses]
[INFO][2023-09-08 19:18:38][ethernet::capture] Available network device: dbus-session (D-Bus session bus): [no interface addresses]
[INFO][2023-09-08 19:18:38][brokers::dot11_broker] Installing WiFi broker thread <0>.
[INFO][2023-09-08 19:18:38][brokers::ethernet_broker] Installing ethernet broker thread <0>.
[INFO][2023-09-08 19:18:38][dot11::capture] Starting WiFi capture on [wlx00c0ca971215]
[INFO][2023-09-08 19:18:38][ethernet::capture] Starting ethernet capture on [enp6s0].
[INFO][2023-09-08 19:18:38][dot11::capture] Starting WiFi capture on [wlx00c0ca956856]
[INFO][2023-09-08 19:18:38][dot11::capture] Temporarily disabling interface [wlx00c0ca971215] ...
[INFO][2023-09-08 19:18:38][dot11::capture] Temporarily disabling interface [wlx00c0ca956856] ...
[ERROR][2023-09-08 19:18:38][dot11::capture] Could not disable device [wlx00c0ca956856]: Could not load device with name [wlx00c0ca956856] for changing state: Could not fetch devices: Could not parse GetIf Netlink response: Missing necessary attributes
[ERROR][2023-09-08 19:18:38][dot11::capture] Could not disable device [wlx00c0ca971215]: Could not load device with name [wlx00c0ca971215] for changing state: Could not fetch devices: Could not parse GetIf Netlink response: Missing necessary attributes
[ERROR][2023-09-08 19:18:38][dot11::channel_hopper] Could not fetch information of device [wlx00c0ca971215]. Not assigning to channels: Could not fetch devices: Could not parse GetIf Netlink response: Missing necessary attributes
[ERROR][2023-09-08 19:18:38][nzyme_tap] WiFi capture [wlx00c0ca971215] disconnected. Retrying in 5 seconds.
[ERROR][2023-09-08 19:18:38][nzyme_tap] WiFi capture [wlx00c0ca956856] disconnected. Retrying in 5 seconds.
[ERROR][2023-09-08 19:18:38][dot11::channel_hopper] Could not fetch information of device [wlx00c0ca956856]. Not assigning to channels: Could not fetch devices: Could not parse GetIf Netlink response: Missing necessary attributes
[INFO][2023-09-08 19:18:38][dot11::channel_hopper] Available adapters and channels: {}
[ERROR][2023-09-08 19:18:38][nzyme_tap] Could not initialize ChannelHopper: WiFi adapter [wlx00c0ca971215] not found.
```

Nothing changed in the `nzyme-tap.conf` yet, right? Attaching mine just in case:

```
[general]
leader_secret = "xxxxxxxxx redacted"
leader_uri = "https://127.0.0.1:22900/"
accept_insecure_certs = true

[ethernet_interfaces.enp6s0]
active = true

[wifi_interfaces.wlx00c0ca956856]
active = true
channels = [1, 6, 11]

[wifi_interfaces.wlx00c0ca971215]
active = true
channels = [2, 3, 4, 5, 7, 8, 9, 10, 12, 13]

[performance]
ethernet_brokers = 1
wifi_brokers = 1
wifi_broker_buffer_capacity = 65535
ethernet_broker_buffer_capacity = 65535

[misc]
training_period_minutes = 5
```

File: tap/src/dot11/nl_skinny_router.rs [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+        }
+        match unsafe { libc::setsockopt(fd, libc::SOL_SOCKET, libc::SO_RCVBUF,  &buf_size as *const _ as *const libc::c_void, std::mem::size_of_val(&buf_size) as u32) } {
+            0 => (),
+            _ => {
+                return Err(SocketError::from(io::Error::last_os_error()));
+            }
+        }
+        socket.bind(Some(0), Groups::empty())?;
+        //let socket = NlSocket::connect(proto, Some(0), Groups::empty())?;
+        //socket.block()?;*/
```
Comment (lennartkoopmann) [inline tap/src/dot11/nl_skinny_router.rs:45]: Can these comments be removed?

File: tap/src/dot11/nl.rs [hunk 2]
```diff
@@ -74,178 +41,151 @@ pub enum InterfaceState {
...
+                                        let freqs_handle = band_attr.get_attr_handle::<Nl80211FrequencyAttr>().map_err(|_| DeError::new("Failed to get WiPhyFreq attribute handle"))?;
+                                        for freq in freqs_handle.iter() {
+                                            // for each element of the array now we get the nested attributes
+                                            let freq_handle = freq.get_attr_handle().map_err(|_| DeError::new("Failed to get WiPhyFreq attribute handle"))?;
+                                            let freq_attr = freq_handle.get_attribute(Nl80211FrequencyAttr::Freq);
+                                            let disabled_attr = freq_handle.get_attribute(Nl80211FrequencyAttr::Disabled);
+
+                                            if freq_attr.is_some() && disabled_attr.is_none() {
+                                                supported_frequencies.push(freq_attr.unwrap().get_payload_as()?);
+                                            }
```
Comment (lennartkoopmann) [inline tap/src/dot11/nl.rs:143]: Same here for the `unwrap()` IMO

File: tap/src/dot11/nl.rs [hunk 3]
```diff
@@ -491,17 +495,25 @@ impl Nl {
...
+//        }
+        /*let _: NlRouterReceiverHandle<GenlId, Nlmsghdr<Rtm, Ifinfomsg>> =
             match self.rt_socket.send(
                 Rtm::Setlink,
                 NlmF::ROOT | NlmF::ECHO,
                 NlPayload::Payload(msg)) {
                 Ok(recv) => recv,
                 Err(e) => { bail!("Could not send request: {}", e); }
-            };
+            };*/
```
Comment (lennartkoopmann) [inline tap/src/dot11/nl.rs:514]: Code comment

File: tap/Cargo.toml [hunk 4]
```diff
@@ -62,6 +62,7 @@ clap = { version = "3.2.15", features = ["derive"] }
 toml = "0.7.3"
 caps = "0.5.5"
 sha2 = "0.10"
-neli = "0.7.0-rc.1"
+neli = "0.7.0-rc2"
 rand = "0.8.5"
-serde_json = "1.0.104"
\ No newline at end of file
+serde_json = "1.0.104"
+libc = "0.2.147"
```
Comment (lennartkoopmann) [inline tap/Cargo.toml:68]: What is `libc` used for? I can't seem to find it in any imports

**Review by lennartkoopmann — CHANGES_REQUESTED** [review]
Just a few minor things for your review. The code works for me!

File: tap/src/dot11/nl.rs [hunk 5]
```diff
@@ -74,178 +41,151 @@ pub enum InterfaceState {
...
-                    break;
-                }
-
-                let element_length: usize = LittleEndian::read_u16(&payload[cursor..cursor+2]) as usize;
-                cursor += 2;
+        interfaces.insert(name.unwrap().clone(), Interface {
+            name: name.unwrap().clone(),
+            phy_index,
+            if_index: if_index.unwrap().clone(),
+        });
```
[the remark this answers is not in the document]
Comment (uskr) [reply tap/src/dot11/nl.rs:105]: There is a really good post about that here: https://blog.burntsushi.net/unwrap/

It is the same recommendation in effect today for the Rust std code and other core parts.

TLDR; It is ok to unwrap if panicking means there is a bug and is not expected behavior. 

I am ok though if we want to limit unwrap usage to tests only and try to match whenever possible and throw an error instead of panicing. Let's discuss this one so we can standardize one way or another.

File: tap/src/dot11/nl.rs (hunk 3 above)
Comment (uskr) [reply tap/src/dot11/nl.rs:514]: Done.

**Comment by uskr** [comment]
Only change still pending is the standardization of the usage of unwrap, let's talk about it on today's call and I will update the PR later today.

File: tap/src/dot11/nl.rs (hunk 2 above)
Comment (uskr) [reply tap/src/dot11/nl.rs:143]: We agreed that unwrap is ok when used in specific situations where a panic would mean a bug.

**Comment by uskr** [comment]
@lennartkoopmann I think this PR is good to go now. Removed all dead code.

**Review by lennartkoopmann — APPROVED** [review]
